core/primitives/src/upgrade_schedule.rs

```diff
@@ -2,233 +2,338 @@ use chrono::{DateTime, NaiveDateTime, ParseError, Utc};
 use near_primitives_core::types::ProtocolVersion;
 use std::env;
 
+#[derive(thiserror::Error, Debug)]
+pub enum ProtocolUpgradeVotingScheduleError {
+    #[error("The final upgrade must be the client protocol version! final version: {0}, client version: {1}")]
+    InvalidFinalUpgrade(ProtocolVersion, ProtocolVersion),
+    #[error("The upgrades must be sorted by datetime!")]
+    InvalidDateTimeOrder,
+    #[error("The upgrades must be sorted and increasing by one!")]
+    InvalidProtocolVersionOrder,
+}
+
 /// Defines the point in time after which validators are expected to vote on the
 /// new protocol version.
-#[derive(Clone, Copy, Debug, PartialEq)]
+#[derive(Clone, Debug, PartialEq)]
 pub struct ProtocolUpgradeVotingSchedule {
-    timestamp: chrono::DateTime<Utc>,
+    /// The schedule is a sorted list of (datetime, version) tuples. The node
+    /// should vote for the highest version that is less than or equal to the
+    /// current time.
+    schedule: Vec<(chrono::DateTime<Utc>, ProtocolVersion)>,
```

Comment:
I went back and forth between vector and some sorted container like BTreeMap. 

I decided to use a vector with a bunch of checks in the Schedule constructor because I do not want to second guess what the release engineer had in mind when setting the schedule. I'd rather verify it's correct and keep it simple.